pkg/dns/dns.go

```diff
@@ -45,25 +56,33 @@ func (s *Server) Shutdown() {
 	}
 }
 
-func newStaticClientConfig(ips []net.IP) (*dns.ClientConfig, error) {
+func newStaticClientConfig(ips []string) (*dns.ClientConfig, error) {
 	s := ``
 	for _, ip := range ips {
-		s += fmt.Sprintf("nameserver %s\n", ip.String())
+		s += fmt.Sprintf("nameserver %s\n", ip)
 	}
 	r := strings.NewReader(s)
 	return dns.ClientConfigFromReader(r)
 }
 
-func newHandler(IPv6 bool, hosts map[string]string) (dns.Handler, error) {
-	cc, err := dns.ClientConfigFromFile("/etc/resolv.conf")
-	if err != nil {
-		fallbackIPs := []net.IP{net.ParseIP("8.8.8.8"), net.ParseIP("1.1.1.1")}
-		logrus.WithError(err).Warnf("failed to detect system DNS, falling back to %v", fallbackIPs)
-		cc, err = newStaticClientConfig(fallbackIPs)
+func newHandler(IPv6 bool, hosts map[string]string, upstreamServers []string) (dns.Handler, error) {
+	var cc *dns.ClientConfig
+	var err error
+	if len(upstreamServers) == 0 {
+		cc, err = dns.ClientConfigFromFile("/etc/resolv.conf")
 		if err != nil {
-			return nil, err
+			logrus.WithError(err).Warnf("failed to detect system DNS, falling back to %v", defaultFallbackIPs)
+			cc, err = newStaticClientConfig(defaultFallbackIPs)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
+
+	cc, err = newStaticClientConfig(upstreamServers)
```

Comment:
Shouldn't this be in an `else` (that is, don't use `upstreamServers` if it's empty)? Otherwise you're clobbering `cc`.

Reply:
Should the tests have caught this?

Reply:
ha, I had a test case for this but I ended up removing it because the docker is always started with `--dns 127.0.0.1` this is due to the fact you cannot edit the `/etc/resolv.conf` once the container started. I ended up removing that test since there is no upstream DNS server so it is pointless to test.